src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -31,6 +31,222 @@ public static bool IsAnyNotNullable(this NullableAnnotation annotation)
+            // Otherwise - result is "oblivious". 
+
+            if (a == NullableAnnotation.NotNullableBasedOnAnalysis || b == NullableAnnotation.NotNullableBasedOnAnalysis)
+            {
+                return NullableAnnotation.NotNullableBasedOnAnalysis;
+            }
+
+            if (a == NullableAnnotation.NotNullable || b == NullableAnnotation.NotNullable)
+            {
+                return NullableAnnotation.NotNullable;
+            }
+
+            Debug.Assert(a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown);
+            return NullableAnnotation.Unknown;
+        }
+
+        /// <summary>
+        /// Meet two nullable annotations from distinct states for the meet (union) operation in flow analysis.
+        /// </summary>
+        public static NullableAnnotation MeetForFlowAnalysisFinally(this NullableAnnotation selfAnnotation, NullableAnnotation otherAnnotation)
+        {
+            if (selfAnnotation == otherAnnotation)
+            {
+                return selfAnnotation;
+            }
+
+            if (selfAnnotation.IsAnyNotNullable() || otherAnnotation.IsAnyNotNullable())
+            {
+                return selfAnnotation == NullableAnnotation.NotNullableBasedOnAnalysis || otherAnnotation == NullableAnnotation.NotNullableBasedOnAnalysis ?
+                            NullableAnnotation.NotNullableBasedOnAnalysis : NullableAnnotation.NotNullable;
+            }
+            else if (selfAnnotation == NullableAnnotation.Unknown || otherAnnotation == NullableAnnotation.Unknown)
+            {
+                return NullableAnnotation.Unknown;
+            }
+            else
+            {
+                return selfAnnotation == NullableAnnotation.NullableBasedOnAnalysis || otherAnnotation == NullableAnnotation.NullableBasedOnAnalysis ?
+                            NullableAnnotation.NullableBasedOnAnalysis : NullableAnnotation.Nullable;
+            }
+        }
+
+        /// <summary>
+        /// Check that two nullable annotations are "compatible", which means they could be the same. Return the
+        /// nullable annotation to be used as a result. Also returns through <paramref name="hadNullabilityMismatch"/>
+        /// whether the caller should report a warning because there was an actual mismatch (e.g. nullable vs non-nullable).
+        /// </summary>
+        public static NullableAnnotation EnsureCompatible<T>(this NullableAnnotation a, NullableAnnotation b, T type, Func<T, bool> isPossiblyNullableReferenceTypeTypeParameter, out bool hadNullabilityMismatch)
+        {
+            hadNullabilityMismatch = false;
+            if (a == b)
+            {
+                return a;
+            }
+
+            if (a.IsAnyNullable() && b.IsAnyNullable())
+            {
+                return NullableAnnotation.Nullable;
+            }
+
+            // If nullability on both sides matches - result is that nullability (trivial cases like these are handled before the switch)
```

Comment:
>the switc [](start = 126, length = 9)

Consider adjusting the comment, the switch is no longer part of the implementation.